vtex-apps/admin-catalog-translations
# Import products backend

**What problem is this solving?**


This PR brings the GraphQL resolvers to allow bulk upload product translations. It will serve the front-end part on PR #13.

There are three new resolvers:

1. Mutation `uploadProductTranslations`

This resolver starts the process to upload the translations. It takes a list of product translations and a locale and return a boolean if there is no error triggering the process. The signature for the product translation is the following:

```ts
interface ProductTranslationUpload {
  id: string
  name?: string
  description?: string
  shortDescription?: string
  title?: string
}
```

*Important:* A translation for a field will only takes effect if the entry for the main language has a value for that field.

All the resolvers logic are declared on the new folder `node/resolvers/product/upload.ts`. Adding this new folder we are preparing the structure so we can split download on its own module as well and rename its resolvers.

The function that serves `uploadProductTranslations` creates and saves on VBase the data structure to hold the request details. In the end, it calls `uploadProductAsync`, which will start translating all the products passed to it. 

**The uploadProductAsync function**

This function loops over all the products, respecting Catalog API throttling limit with a `pacer` function after every loop. 

The translation is triggered calling a new method `translateProduct` on `catalogGQL` client. It sends a mutation to `catalog-graphql` app.

The function also defines the indexes that match the progress breakpoints 25%, 50%, and 75%. Those will be used to update the request status.

Inside the loop, when it is in one of the breakpoints described above, the loop will resolve the promises collected until that point (this way we can raise a error if any of the calls have failed) and will update VBase with the translation progress.

2. Query `productTranslationsUploadRequests`

This query returns the id of all the requests made to upload product translations.

3. Query `translationUploadRequestInfo`

This query returns the information and status of a request given a request id. The returned object has the following signature:

```ts
interface UploadRequest {
  requestId: string
  translatedBy: string
  createdAt: String
  estimatedTime: number
  error?: boolean // when an if there is an error during the upload process.
  progress?: number // a integer representing the %. It can be ~25%, ~50%, ~75% or 100%.
  locale?: string
}
```

**How should this be manually tested?**

The new resolvers are available in [this workspace](https://catalogimportbackend--powerplanet.myvtex.com/admin/graphql-ide).

Example to trigger an upload:

```gql
mutation {
  uploadProductTranslations(
    locale: "fr-FR",
    products: [
      {
          id: "27412",
    	  name: "iPhone 7 32 Go Rose Gold updatedName",
          description: "New description for 27412"
          title: "New title for 27412"
      },
      {
          id: "27413",
          name: "iPhone 7 32 Go Argent updatedName",
          description: "New description for 27413"
          title: "New title for 27413"
      },
      {
          id: "27414",
    	  name: "iPhone 7 32 Go Oro updatedName",
          description: "New description for 27414",
          title: "New title for 27414"
      },
    ]
  )
}
```

After sending an upload, it is possible to get the request id using the query `productTranslationsUploadRequests`. The most recent request made will always be the first string in the array.

To see the request information, use the query `translationUploadRequestInfo`, passing the request id to it. E.g.:

```gql
translationUploadRequestInfo(requestId: "40c8bb96c0654570892ea214484d8837") {
    requestId
    translatedBy
    createdAt
    estimatedTime
    error
    progress
    locale
  }
```
